RavenDB-14516: Fixed - revisions(array) and revisions(property),revisions(property) with cache to sync/async API.

omerlewitz opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-14516

Commit:
RavenDB-14516 - infrastructure

Commit:
RavenDB-14516 - added the given functionality [WIP]

Commit:
RavenDB-14516 changing base code[WIP]

omerlewitz committed:
RavenDB-14516 WIP

Commit:
RavenDB-14516 addedArrayTest functionality WIP

Commit:
RavenDB-14516 - WIP: cache (need to verify round trip to the server)
                raw query new logic.

omerlewitz commented:
WIP regarding checks

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -110,7 +111,12 @@ public async Task<ServerNode> GetCurrentSessionNode()
+        /// <summary>
+        /// Translate between an CV and its associated entity
+        /// </summary>
+        internal readonly Dictionary<string, DocumentInfo> IncludedRevisionByChangeVectors = new Dictionary<string, DocumentInfo>(StringComparer.OrdinalIgnoreCase);
```
This will almost always be null, let's initialize this only when it is actually used.

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1420,7 +1429,35 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
+            {
+                includes.GetPropertyByIndex(i, ref propertyDetails);
+
+                if (propertyDetails.Value == null)
```
```suggestion
                if (propertyDetails.Value is BlittableJsonReaderObject json == false)
```

omerlewitz replied:
done

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1574,6 +1611,17 @@ private void RegisterCountersForDocument(string id, bool gotAll, BlittableJsonRe
+                if (string.IsNullOrEmpty(cv) == false)
+                    continue;
+
+                SetGotAllCountersForDocument(cv);
```
Why are you calling counters here?

omerlewitz replied:
Redundant

ayende commented on src/Raven.Server/Documents/Includes/IncludeRevisionsCommand.cs:
```diff
@@ -0,0 +1,51 @@
+
+            foreach (var fieldName in _revisionsBySourcePath)
+            {
+                if ( (fieldName != string.Empty && document.Data.TryGet(fieldName, out revisionCV)) == false)
```
That is for a single value, no? How do you handle arrays here?

omerlewitz replied:
Resolved

Commit:
RavenDB-14516: [WIP] functionality working excluding Jint

Commit:
RavenDB-15152 - dealing with IncludeRevisionComman null issues

Commit:
RavenDB-14516: Fixed - revision(array) and revision(property),revision(property)
with cache to sync/async API.

Commit:
RavenDB-14516: support for change vectors array [WIP]

Commit:
RavenDB-14516: Resolving conflict regarding lazy revisions. Fixing NRE causing test failures

Commit:
RavenDB-14516: Missing `FirstOrDefault()`

ayende commented on src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs:
```diff
@@ -11,7 +11,7 @@ namespace Raven.Client.Documents.Commands
     public class GetRevisionsCommand : RavenCommand<BlittableArrayResult>
     {
-        private readonly string _changeVector;
+        internal readonly string _changeVector;
```
Why change the visibility?

omerlewitz replied:
I'm using `_changeVector` inside GetRevisionOperation to verify that a given cv is within the Cache. I didn't want to change the implementation of `GetRequestQueryString` as well as one of the signatures inside `GetRevisionsCommand`. In other words, I could work with ChangeVectors and that's it. I thought the best solution is just to change the visibility  like in `changeVectors property`

ayende replied:
In that case, expose that to public (we already have `ChangeVectors`). Accessing the internal state in this manner isn't something we want to encourage.

ayende commented on src/Raven.Client/Documents/Session/DocumentSessionRevisions.cs:
```diff
@@ -122,6 +129,7 @@ public long GetCountFor(string id)
         {
             var operation = new GetRevisionsCountOperation(id);
             var command = operation.CreateRequest();
+            if (command == null) return command.Result;
```
This is wrong, will cause NRE

omerlewitz replied:
I'm looking into it and I wonder why I changed it. It doesn't make any sense to me whatsoever.
there's no need to caching over here because it's returning the number of revisions and that's it.
the only thing I will change here is the IncrementNumberOfRequests. sounds good?

ayende commented on src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs:
```diff
@@ -34,7 +34,8 @@ public async Task<List<T>> GetForAsync<T>(string id, int start = 0, int pageSize
             {
                 var operation = new GetRevisionOperation(Session, id, start, pageSize);
                 var command = operation.CreateRequest();
-                await RequestExecutor.ExecuteAsync(command, Context, SessionInfo, token).ConfigureAwait(false);
```
Can this be null?

omerlewitz replied:
I was missing return ..

ayende commented on src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs:
```diff
@@ -95,7 +104,9 @@ public async Task<long> GetCountForAsync(string id, CancellationToken token = de
                 var operation = new GetRevisionsCountOperation(id);
                 var command = operation.CreateRequest();
-                await RequestExecutor.ExecuteAsync(command, Context, sessionInfo: SessionInfo, token: token).ConfigureAwait(false);
+                if (command == null) return command.Result;
```
This is wrong, NRE

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1420,7 +1429,36 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
+                    continue;
+                
+                if (newDocumentInfo.Metadata.TryGet(Constants.Documents.Metadata.ChangeVector,out string cv))
+                    
```
This is really confusing with the extra line break here

ayende commented on src/Raven.Client/Documents/Session/Operations/QueryOperation.cs:
```diff
@@ -187,6 +187,12 @@ private void CompleteInternal<T>(QueryResult queryResult, Action<int, T> addToRe
+                if (queryResult.RevisionIncludes != null)
+                {
+                    _session.RegisterRevisionIncludes(
+                        queryResult.RevisionIncludes);
```
Single line

ayende commented on src/Raven.Server/Documents/Includes/IncludeRevisionsCommand.cs:
```diff
@@ -0,0 +1,78 @@
+            {
+                if (document.Data.TryGet(fieldName, out object singleOrMultipleCv) == false  )
+                {
+                    throw new InvalidOperationException($"Cannot include revisions for related document '{document.Id}', " +
+                                                        $"document {document.Id} doesn't have a field named '{fieldName}'. ");
```
Is this something that we do for includes as well?
If the value doesn't exists, I think we ignore that, no?

omerlewitz replied:
Regarding `Documents` we just ignore it although in `IncludecounterCommand` as well as `IncludeTimeSeriesCommand` we throw an exception like in the above.

ppekrol replied:
Sth that we definitely need to revisit. I do not see a reason to throw here? Should behave as similar as possible to IncludeDocuments I think?

ayende commented on src/Raven.Server/Documents/Includes/IncludeRevisionsCommand.cs:
```diff
@@ -0,0 +1,78 @@
+                        }
+                        break;
+                    }
+                    case LazyStringValue lazyStringValue:
```
Can this be LazyCompressedString?

omerlewitz replied:
I was trying to patch cv to a given document which clearly did not work.
The second option was to take into consideration `LazyCompressedString` so I added another case

ayende commented on src/Raven.Server/Documents/Patch/ScriptRunner.cs:
```diff
@@ -1752,7 +1784,26 @@ private JsValue CmpXchangeInternal(string key)
                 }
             }
+            
+            private JsValue LoadRevisionInternal(IEnumerable<JsValue> cvs)
```
who is using this?

omerlewitz replied:
Redundant

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +582,52 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                            break;
+                        }
+
+                        if(split[0] == vt.Value)
```
See the warnings here

omerlewitz replied:
reduce nesting?

omerlewitz replied:
just noticed the warning in github. I will fix it..

ayende commented on src/Raven.Server/Documents/Queries/QueryValidator.cs:
```diff
@@ -119,7 +119,25 @@ public static void ValidateIncludeCounter(List<QueryExpression> arguments, strin
+                switch (arguments[i])
+                {
+                    case FieldExpression :
+                    case ValueExpression :
```
can this be a parameter?

omerlewitz replied:
`ParameterExpression` is a type of ValueExpression. 
https://github.com/ravendb/ravendb/blob/v5.1/src/Raven.Server/Documents/Queries/AST/ValueTokenType.cs

ayende commented on test/FastTests/Server/Documents/Indexing/BasicAnalyzers.cs:
```diff
@@ -238,8 +238,8 @@ public override (ICollection<string> Static, ICollection<string> Dynamic) GetEnt
             throw new NotImplementedException();
```
What did you change in the `src/Raven.Server/ServerWide/OperationCancelToken.cs ` file?

omerlewitz replied:
breaking line have been removed

Commit:
RavenDB-14516: Changes discussion

ayende commented on src/Raven.Client/Documents/Session/DocumentSessionRevisions.cs:
```diff
@@ -124,12 +124,12 @@ public T Get<T>(string id, DateTime date)
             var command = operation.CreateRequest();
-            if (command == null) return command.Result;
+            SessionInfo?.IncrementRequestCount();
             RequestExecutor.Execute(command, Context, sessionInfo: SessionInfo);
```
Can `command` be null here?

omerlewitz replied:
nope

ayende commented on src/Raven.Client/Documents/Session/DocumentSessionRevisionsAsync.cs:
```diff
@@ -104,7 +105,6 @@ public async Task<long> GetCountForAsync(string id, CancellationToken token = de
             {
                 var operation = new GetRevisionsCountOperation(id);
                 var command = operation.CreateRequest();
-                if (command == null) return command.Result;
```
Can command be null here?

omerlewitz replied:
nope

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -612,15 +612,12 @@ private void AddToRevisionsInclude(RevisionIncludeField revisionIncludes, Method
-                        throw new InvalidOperationException($"Cannot include revisions for related Expression '{vt}', " + 
-                                                            $"Parent alias is different than include alias '{Query.From.Alias.Value}'" +
-                                                            $" compare to '{split[0]}';. ");
+                        if (split[0] != vt.Value)
```
See the warning

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -110,7 +111,12 @@ public async Task<ServerNode> GetCurrentSessionNode()
+        /// <summary>
+        /// Translate between an CV and its associated entity
+        /// </summary>
+        internal  Dictionary<string, DocumentInfo> IncludedRevisionByChangeVectors;
```
typo: IncludedRevisionsByChangeVectors ?

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -133,9 +139,12 @@ public async Task<ServerNode> GetCurrentSessionNode()
 
         protected readonly DocumentStoreBase _documentStore;
 
+
```
There is a lot of additional empty lines, please check the out

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1408,10 +1417,10 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
                 includes.GetPropertyByIndex(i, ref propertyDetails);
 
-                if (propertyDetails.Value == null)
+                if (propertyDetails.Value is BlittableJsonReaderObject json == false)
```
Can we have sth else here?

omerlewitz replied:
as far as I see no, although Oren suggest this change

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1420,7 +1429,35 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
+                var newDocumentInfo = DocumentInfo.GetNewDocumentInfo(json);
+                if (newDocumentInfo.Metadata.TryGetConflict(out var conflict) && conflict)
+                    continue;
+                if (newDocumentInfo.Metadata.TryGet(Constants.Documents.Metadata.ChangeVector,out string cv))
```
Is there a chance that there will be no CV? Maybe we should throw then?

omerlewitz replied:
I think the best idea is to be consistent as you said. compare to the `include document` we return rather than throwing an error.
https://github.com/ravendb/ravendb/pull/12242/files#r648209360

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -2072,6 +2109,21 @@ private object DeserializeFromTransformer(Type entityType, string id, BlittableJ
             return JsonConverter.FromBlittable(entityType, ref document, id, trackEntity);
         }
+        
+        public bool CheckIfChangeVectorAlreadyIncluded(IEnumerable<string> changeVectors)
```
internal?

ppekrol commented on src/Raven.Server/Documents/Includes/IncludeRevisionsCommand.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Raven.Server.Documents.Includes
+{
+    public class IncludeRevisionsCommand
```
better to switch this to internal

omerlewitz replied:
When I change it I'm getting accessibility issues in several places  
![image](https://user-images.githubusercontent.com/39852633/121806901-74955a00-cc5a-11eb-908e-7f547fe55333.png)

ppekrol commented on src/Raven.Server/Documents/Queries/Dynamic/CollectionQueryRunner.cs:
```diff
@@ -165,9 +168,17 @@ private async ValueTask ExecuteCollectionQueryAsync(QueryResultServerSide<Docume
                 }
 
-                IncludeCountersCommand includeCountersCommand = null;
+                IncludeCountersCommand   includeCountersCommand   = null;
```
formatting

ppekrol commented on src/Raven.Server/Documents/Queries/GraphQueryRunner.cs:
```diff
@@ -140,9 +141,10 @@ public async Task WriteDetailedQueryResult(IndexQueryServerSide indexQuery, Quer
+                        idc   = new IncludeDocumentsCommand(Database.DocumentsStorage, queryContext.Documents, query.Metadata.Includes, fieldsToFetch.IsProjection);
                         icevc = IncludeCompareExchangeValuesCommand.ExternalScope(queryContext, query.Metadata.CompareExchangeValueIncludes);
-
+                        irc    = new IncludeRevisionsCommand(database:Database, context:queryContext.Documents, pathsForRevisionsInDocuments: query.Metadata.RevisionIncludes?.Revisions);
```
formatting

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +582,49 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+            } 
+        }
+        
+        private static ExplanationField CreateExplanationField(MethodExpression expression)
```
formatting

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +582,49 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                {
+                    case FieldExpression fe:
+                    {
+                        if (Query.From.Alias?.Value.Equals(fe.Compound[0].Value) == false)
```
Talk with @aviv86 how to handle this. I think we do have a dedicated set of methods for extraction of those.

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +582,49 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                                                                $"Parent alias is different than include alias '{Query.From.Alias?.Value}'" +
+                                                                $" compare to '{fe.Compound[0].Value}';. ");
+                      
+                        if(string.IsNullOrEmpty(fe.FieldValueWithoutAlias) == false)
```
Talk with @aviv86 how to handle this. I think we do have a dedicated set of methods for extraction of those.

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +582,49 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                        else revisionIncludes.AddRevision(fe.FieldValue);
+                        break;
+                    }
+                    case ValueExpression {Value: ValueTokenType.Parameter} ve:
```
IIRC QueryBuilder does have some methods that extract values appropriately with error handling.

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -749,8 +805,9 @@ private void AddToCounterIncludes(CounterIncludesField counterIncludes, MethodEx
                                                 $"but got `{parameterValue.Value}` of type `{parameterValue.Type}`", QueryText, parameters);
 
             counterIncludes.AddCounter(parameterValue.Value.ToString(), sourcePath);
+            counterIncludes.AddCounter(parameterValue.Value.ToString(), sourcePath);
```
duplicate?

omerlewitz replied:
yes

Commit:
RavenDB-14516: [WIP] resolved conversation

Commit:
RavenDB - 14516 WIP

Commit:
RavenDB-14516: resolved converstation for review

omerlewitz reviewed (commented):
@ppekrol  I forgot to comment about https://github.com/ravendb/ravendb/pull/12242#discussion_r648213079 
Aviv told me it fits the logic and it's ok. The second option is to manipulate `fe.FieldValue`

ppekrol commented on src/Raven.Client/Documents/Commands/GetRevisionsCommand.cs:
```diff
@@ -11,7 +11,7 @@ namespace Raven.Client.Documents.Commands
     public class GetRevisionsCommand : RavenCommand<BlittableArrayResult>
     {
-        private readonly string _changeVector;
+        public readonly string ChangeVector;
```
why it is public? does it have to? Maybe internal?

ppekrol commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -110,7 +111,12 @@ public async Task<ServerNode> GetCurrentSessionNode()
+        /// <summary>
+        /// Translate between an CV and its associated entity
+        /// </summary>
+        internal  Dictionary<string, DocumentInfo> IncludeRevisionsByChangeVector;
```
formatting

omerlewitz replied:
done

ppekrol commented on test/SlowTests/Tests/Linq/CanQueryAndIncludeRevisions.cs:
```diff
@@ -0,0 +1,731 @@
+                using (var session = store.OpenAsyncSession())
+                {
+                    var query = await session.Advanced
+                        .AsyncRawQuery<User>("from Users as u include revisions($p0),revisions($p1),revisions($p2)")
```
We also need to add include methods to IIncludeBuilder in the Client API

Commit:
RavenDB-14516 support for jint with array

Commit:
adding revisions in Index.cs

Commit:
Added IncludeBuilder Load/Query functionality. I need to verify regression tests

Commit:
Waiting for aviv's review

Commit:
RavenDB-14516: Redundant code, Indentation, resolving `command.Result` initialization

topalach commented:
test this please

topalach commented:
> test this please

This is a comment that will reschedule a jenkins PR test build.

Commit:
RavenDB-14516: Fixed and rebased

topalach commented:
test this please

ayende commented on test/SlowTests/Tests/Linq/CanQueryAndIncludeRevisions.cs:
```diff
@@ -0,0 +1,1982 @@
+                        .RawQuery<User>(
+                            @"
+declare function Foo(i) {
+    includes.revisions(i.ChangeVector)
```
How do I get revisions before from JS?

ayende commented on test/SlowTests/Tests/Linq/CanQueryAndIncludeRevisions.cs:
```diff
@@ -0,0 +1,1982 @@
+                    var changeVector = metadatas.First().GetString(Constants.Documents.Metadata.ChangeVector);
+                    var getRevisionsBefore = DateTime.UtcNow;
+                    var query =  session.Advanced
+                        .RawQuery<User>("from Users as u include revisions($p0)")
```
In here, we have basically:

* `revisions(string)` which is a property in the document
* `revisions(date)` which is revision before, right?

I think it would be better to have different names.

ayende commented on test/SlowTests/Tests/Linq/CanQueryAndIncludeRevisions.cs:
```diff
@@ -0,0 +1,1982 @@
+                {
+                    var ex = Assert.Throws<InvalidOperationException>(()=> 
+                                 session.Load<User>(id,builder => builder
+                                .IncludeRevisions(x => x.ChangeVectors)
```
Why can't I do this here in Load?

ayende commented on test/SlowTests/Tests/Linq/CanQueryAndIncludeRevisions.cs:
```diff
@@ -0,0 +1,1982 @@
+                {
+                    var query = session.Advanced
+                        .RawQuery<User>("from Users as u include revisions($p0, $p1, $p2)")
+                        .AddParameter("p0", "u.FirstRevision")
+                        .AddParameter("p1", "u.SecondRevision")
+                        .AddParameter("p2", "u.ThirdRevision")
+                        .ToList();
```
This is _super_ strange. You are passing the `u` as the alias here? But that is almost never how this would actually work.

ayende commented on test/SlowTests/Tests/Linq/CanQueryAndIncludeRevisions.cs:
```diff
@@ -0,0 +1,1982 @@
+                {
+                    
+                    var error =  Assert.ThrowsAny<RavenException>( () =>  session.Advanced
+                        .RawQuery<User>("from Users as u include revisions($p0, $p1, $p2)")
+                        .AddParameter("p0", "u.FirstRevision")
+                        .AddParameter("p1", "u.SecondRevision")
+                        .AddParameter("p2", "x.ThirdRevision")
```
@ppekrol How do you read this? I would lean toward just killing passing this via parameters entirely, given the alias situation

ppekrol replied:
I do not recall us having similar syntax elsewhere. Maybe it just works out of the box? @omerlewitz you had to add support for this explicitly? If yes, probably we should drop it. Or maybe there is some case where this would be useful?

ppekrol commented on src/Raven.Client/Documents/Session/Loaders/IncludeBuilder.cs:
```diff
@@ -90,8 +101,11 @@ public interface IQueryIncludeBuilder<T> : IIncludeBuilder<T, IQueryIncludeBuild
+        
         IQueryIncludeBuilder<T> IncludeTimeSeries(Expression<Func<T, string>> path, string name, DateTime from, DateTime to);
+        
+        //IQueryIncludeBuilder<T> IncludeRevisions(DateTime dateTime);
```
cleanup

omerlewitz replied:
thanks~!

ppekrol commented on src/Raven.Client/Documents/Session/Operations/GetRevisionOperation.cs:
```diff
@@ -46,7 +45,12 @@ public GetRevisionOperation(InMemoryDocumentSessionOperations session, IEnumerab
+            if (_command.ChangeVectors is not null)
+                return _session.CheckIfChangeVectorAlreadyIncluded(_command.ChangeVectors) ? null : _command ;
+            
+            if (_command.ChangeVector is not null)
```
duplicated code?

omerlewitz replied:
Nope, it is because there are two commands -  one for change vector and one for change vectors

ppekrol commented on src/Raven.Client/Documents/Session/SessionInfo.cs:
```diff
@@ -28,6 +28,11 @@ public int SessionId
             }
         }
+        
+        public void IncrementRequestCount()
```
internal

ppekrol commented on src/Raven.Client/Documents/Session/Tokens/RevisionIncludesToken.cs:
```diff
@@ -0,0 +1,56 @@
+
+            }
+
+            else if(string.IsNullOrEmpty(_changeVector) == false)
```
changeVector?

omerlewitz replied:
in both cases, it creates RevisionToken with a given change vector each time so there are two overloads but it is inserted into the same collection

```
        private void IncludeRevisionsByChangeVector(string changeVector)
        {
            if (string.IsNullOrWhiteSpace(changeVector))
                throw new ArgumentNullException(nameof(changeVector));
            
            RevisionsToIncludeByChangeVector ??= new HashSet<string>();
            RevisionsToIncludeByChangeVector.Add(changeVector);
        }
        
        private void IncludeRevisionsByChangeVectors(IEnumerable<string> changeVectors)
        {
            RevisionsToIncludeByChangeVector ??= new HashSet<string>();
            foreach (var changeVector in changeVectors)
            {
                RevisionsToIncludeByChangeVector.Add(changeVector);
            }
        }
```

Commit:
Changes in dateTime cache

Commit:
WIP

Commit:
WIP

Commit:
RavenDB-14516:
Uniting the results to be in one object.
dateTime cache is by UTC, including revision with local DateTime will convet it to UTC

ayende commented on src/Raven.Client/Documents/Session/AbstractDocumentQuery.Includes.cs:
```diff
@@ -110,5 +122,20 @@ private void IncludeTimeSeries(string alias, Dictionary<string, HashSet<Abstract
             }
         }
+        
+        private void IncludeRevisions(DateTime? dateTime)
```
This cannot actually accept nullable, so let's remove it.
```suggestion
        private void IncludeRevisions(DateTime dateTime)
```

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1443,10 +1456,10 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
                     continue;
 
-                var json = (BlittableJsonReaderObject)propertyDetails.Value;
+                json = (BlittableJsonReaderObject)propertyDetails.Value;
```
You can delete this line now, no?

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1456,6 +1469,38 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
+                if (obj is BlittableJsonReaderObject json == false)
+                    continue;
+                
+                json = ((BlittableJsonReaderObject)obj);
```
Can delete this line, no?

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -1456,6 +1469,38 @@ internal void RegisterIncludes(BlittableJsonReaderObject includes)
+                
+                IncludeRevisionsByChangeVector[changeVector] = DocumentInfo.GetNewDocumentInfo(revision);
+                
+                if(dateTime != default && string.IsNullOrWhiteSpace(id) == false)
```
Missing `{}`

ayende commented on src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs:
```diff
@@ -2107,7 +2152,35 @@ private object DeserializeFromTransformer(Type entityType, string id, BlittableJ
+            
+            foreach (var cv in changeVectors)
+            {
+                if (IncludeRevisionsByChangeVector.TryGetValue(cv, out DocumentInfo documentInfo) == false )
+                    return false;
+                
+                if ((documentInfo.Entity == null) && (documentInfo.Document == null))
+                    return false;
```
This logic seems suspicious. shouldn't you return `true` here when the value is found?

ayende commented on src/Raven.Client/Documents/Session/Loaders/IncludeBuilder.cs:
```diff
@@ -123,6 +133,8 @@ internal HashSet<string> CountersToInclude
                 return value.CountersToInclude;
             }
         }
+        //TODO:delete
+        //internal HashSet<string> RevisionToInclude => RevisionsToIncludeByChangeVector;
 
```
Delete

ayende commented on src/Raven.Client/Documents/Session/Loaders/IncludeBuilder.cs:
```diff
@@ -525,6 +538,27 @@ private void IncludeCountersWithAlias(Expression<Func<T, string>> path, string[]
+            if (dateTime == null)
+                return;
+            
+            if(dateTime == default(DateTime))
+                throw new ArgumentNullException(nameof(dateTime));
+    
```
What is this scenario? why ?

ayende commented on src/Raven.Client/Documents/Session/Loaders/IncludeBuilder.cs:
```diff
@@ -525,6 +538,27 @@ private void IncludeCountersWithAlias(Expression<Func<T, string>> path, string[]
+            if(dateTime == default(DateTime))
+                throw new ArgumentNullException(nameof(dateTime));
+    
+            RevisionsToIncludeByDateTime ??= new DateTime();
```
Why?

omerlewitz replied:
In the JINT side it will always initialize date time for the first time

ayende commented on src/Raven.Client/Documents/Session/Tokens/RevisionIncludesToken.cs:
```diff
@@ -0,0 +1,49 @@
+
+            }
+
+            else if(string.IsNullOrEmpty(_path) == false)
+            {
+                writer.Append($"{_path}");
+            }
```
I wonder if we should handle escaping here?

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,55 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                        if (ParquetTransformedItems.TryParseDate(fe.FieldValueWithoutAlias, out dateTimeOffset) == false && fe.Compound.Count >= 2)
+                            throw new InvalidOperationException($"Alias is not supported `include revisions(..)`.");
+                        
+                        if (dateTimeOffset != null)
```
Can remove this `if`

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,55 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                        if (Query.From.Alias!= null)
+                            throw new InvalidOperationException($"Alias is not supported `include revisions(..)`.");
+
+                        if (ParquetTransformedItems.TryParseDate(fe.FieldValueWithoutAlias, out dateTimeOffset) == false && fe.Compound.Count >= 2)
```
`ParquetTransformedItems` ??

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,55 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                              
+                            string[] split = valueType.Value.ToString()?.Split('.');
+                            
+                            if (ParquetTransformedItems.TryParseDate(valueType.Value.ToString(), out dateTimeOffset) == false && split.Length >= 2)
```
`ParquetTransformedItems` ??

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,55 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                            if (Query.From.Alias != null)
+                                throw new InvalidOperationException($"Alias is not supported `include revisions(..)`.");
+                              
+                            string[] split = valueType.Value.ToString()?.Split('.');
```
why are you checking this?

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,55 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                            string[] split = valueType.Value.ToString()?.Split('.');
+                            
+                            if (ParquetTransformedItems.TryParseDate(valueType.Value.ToString(), out dateTimeOffset) == false && split.Length >= 2)
+                                throw new InvalidOperationException($"Alias is not supported `include revisions(..)`.");
```
And if the key they are using is `Foo.Bar'` ?

ayende commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,55 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                            if (ParquetTransformedItems.TryParseDate(valueType.Value.ToString(), out dateTimeOffset) == false && split.Length >= 2)
+                                throw new InvalidOperationException($"Alias is not supported `include revisions(..)`.");
+
+                            if (dateTimeOffset != null)
```
if can be removed

ayende commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -1467,6 +1478,69 @@ public static async Task WriteIncludesAsync(this AsyncBlittableJsonTextWriter wr
+
+                        writer.WriteStartObject();
+                    
+                        writer.WritePropertyName("Id");
```
use `nameof` here

Commit:
RavenDB-14516: Oren's remarks, added RevisionIncludeResult class for nameof

ppekrol commented on src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs:
```diff
@@ -158,6 +169,19 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
                 }
             }
 
+            if (_revisionsIncludeByChangeVector != null && _revisionsIncludeByChangeVector.Any())
```
you can skip `.Any()` check, because on IEnumerable this and foreach will cause a double enumeration. Just leave `foreach` in this case.

omerlewitz replied:
Tnx!

ppekrol commented on src/Raven.Client/Documents/Commands/GetDocumentsCommand.cs:
```diff
@@ -158,6 +169,19 @@ public override HttpRequestMessage CreateRequest(JsonOperationContext ctx, Serve
+            
+            if (_revisionsIncludeByDateTime != null)
+            {
+                pathBuilder.Append("&revisionBeforeDateTime=").Append(Uri.EscapeDataString(_revisionsIncludeByDateTime.Value.GetDefaultRavenFormat()));
```
`revisionsBefore` instead of `revisionBeforeDateTime`

ppekrol commented on src/Raven.Client/Documents/Session/Loaders/IncludeBuilder.cs:
```diff
@@ -53,6 +53,14 @@ public interface IAbstractTimeSeriesIncludeBuilder<T, out TBuilder>
+    {
+        TBuilder IncludeRevisions(Expression<Func<T, string>> path);
+        TBuilder IncludeRevisions(Expression<Func<T, IEnumerable<string>>> path);
+        TBuilder IncludeRevisions(DateTime dateTime);
```
rename maybe `dateTime` to `before` ?

ppekrol commented on src/Raven.Server/Documents/Queries/QueryMetadata.cs:
```diff
@@ -567,8 +584,42 @@ void AddInclude(QueryExpression include, string path, ref List<string> listToAdd
+                    case ValueExpression ve:
+                        foreach ((object value, _)  in QueryBuilder.GetValues(Query, this, parameters, ve))
+                        {
+                            string? path = value.ToString();
```
Remove the nullable annotation, it produces a warning

omerlewitz replied:
it is strange, In Rider, it doesn't show any error whatsoever

ppekrol commented on src/Raven.Server/Json/BlittableJsonTextWriterExtensions.cs:
```diff
@@ -1467,6 +1479,69 @@ public static async Task WriteIncludesAsync(this AsyncBlittableJsonTextWriter wr
+                        WriteDocument(writer, context, metadataOnly: false, document: doc);
+                    
+                        writer.WriteEndObject();
+                    }
```
don't we want to execute `MaybeFlushAsync` here?

Commit:
RavenDB-14516: Pawel's comments

Commit:
RavenDB-14516

Wrong location of MabyFlushAsync
